app/database_etl/test_evaluation_handler/test_eval_link_handler.py

```diff
@@ -0,0 +1,100 @@
+    test_name, manufacturers, target_isotypes = row['test_name'], row['test_manufacturer'], row['isotypes']
+
+    # CRITERIA A – test name, manufacturer, and target isotype
+    # Get all rows matching test name
+    matches = test_eval_df[test_eval_df['TestName'] == test_name]
+    if matches.empty: return [None]*2
+
+    if manufacturers[0] == 'Zhejiang Orient Gene (Biotech Co LTD':
+        manufacturers = ['Zhejiang Orient Gene (Biotech Co LTD China)']
+
+    temp = pd.DataFrame()
+    # Get all rows matching any name in manufacturer list
+    for manufacturer in manufacturers:
+        if manufacturer in manufacturer_mapping: manufacturer = manufacturer_mapping[manufacturer]
+        manu = matches[matches['Manufacturer'].str.contains(f"(?i){manufacturer}")]
+        temp = temp.append(manu)
+
+    matches = temp
+    # Check that targets are exact match
+    for i, ev in matches.iterrows():
+        if not set(ev['Target']) == set(target_isotypes):
+            matches.drop(i)
+
+    if matches.empty: return [None]*2
+    if matches.shape[0] == 1: return [matches.iloc[0]['unique_id'], matches.iloc[0]['DOI']]
+
+    # CRITERIA B – sample specimen type
+    specimen_type = row['specimen_type']
+    new_matches = matches[matches['IndexSampleType'] == specimen_type]
+
+    if not new_matches.empty:
+        matches = new_matches
+        if matches.shape[0] == 1: return [matches.iloc[0]['unique_id'], matches.iloc[0]['DOI']]
+
+    # CRITERIA C – prioritized reference specimen type
+    rankings = {'Lower respiratory specimen': 1, 'Sputum': 3, 'Bronchial aspirate': 3,
+                'Respiratory specimen': 4, 'Upper respiratory specimen': 4, 'Nasal mid-turbinate swabs': 5,
+                'Nasopharyngeal swabs': 7, 'Swab': 7, 'Serum': 8, 'Pharyngeal swabs': 10, 'Plasma': 11, 'Oropharyngeal swabs': 12,
+                'Whole blood': 13, 'Urine': 14, 'Not available': 20, 'not reported': 20, 'Not reported': 20}
+
+    min_rank, top_matches = 100, pd.DataFrame()
+    for i, ev in matches.iterrows():
+        if pd.isna(ev['ReferenceSampleType']):
+            if top_matches.empty or min_rank == 20:
+                min_rank = 20
+                top_matches = top_matches.append(ev)
+                continue
+        reference_specimen_types = ev['ReferenceSampleType'].split('; ')
+        for ref in reference_specimen_types:
+            rank = rankings[ref]
+            if rank > min_rank: continue
+            if rank == min_rank: top_matches = top_matches.append(ev)  # Same rank, so append
+            else:  # rank < min_rank
+                min_rank, top_matches = rank, ev.to_frame().T
+
+    if not top_matches.empty:
+        matches = top_matches
+        if matches.shape[0] == 1: return [matches.iloc[0]['unique_id'], matches.iloc[0]['DOI']]
+
+    # CRITERIA D – largest sample size
+    largest_sample_size, top_match = 0, pd.DataFrame()
```

Comment:
You're just trying to get the row in the matches df with the highest SampleSize here right? You could probably just do this:

```top_match = matches[matches['SampleSize'] ==matches['SampleSize'].max()]```

You'll need to dbl check my syntax here but you get the idea

Reply:
idxmax could be better suited for the inside of the square brackets here- see https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

Reply:
Okay with idxmax if the use case is that we for sure want to return one row. But the approach I provided can also be used if we want to return all rows that feature a max value for some column, so I think it's a bit more extensible there. 

Either or is a good option for me